code/game/machinery/kitchen/microwave.dm

```diff
@@ -255,167 +283,146 @@
-			reagents.trans_to_holder(S.reagents, total/cooked_items.len)
-
-		for (var/obj/item/weapon/reagent_containers/food/snacks/S in contents)
-			S.cook()
-
-		dispose(0) //clear out anything left
-		stop()
+/obj/machinery/microwave/proc/update_cook_time(var/ct = 200)
+	RefreshParts()
+	return (ct / cooking_power)
+
+/obj/machinery/microwave/proc/finish_cooking()
+	var/result = recipe.result
+	var/valid = TRUE
+	var/list/cooked_items = list()
+	var/obj/temp = new /obj(src) //To prevent infinite loops, all results will be moved into a temporary location so they're not considered as inputs for other recipes
+	while(valid)
+		var/list/things = list()
+		things.Add(recipe.make_food(src))
+		cooked_items += things
+		//Move cooked things to the buffer so they're not considered as ingredients
+		for (var/atom/movable/AM in things)
+			AM.forceMove(temp)
+
+		valid = FALSE
+		recipe = select_recipe(RECIPE_LIST(appliancetype),src)
+		if (recipe && recipe.result == result)
+			sleep(2)
+			valid = TRUE
+
+	for (var/r in cooked_items)
+		var/atom/movable/R = r
+		R.forceMove(src) //Move everything from the buffer back to the container
+
+	QDEL_NULL(temp)//Delete buffer object
+
+	//Any leftover reagents are divided amongst the foods
+	var/total = reagents.total_volume
+	for (var/obj/item/weapon/reagent_containers/food/snacks/S in cooked_items)
+		reagents.trans_to_holder(S.reagents, total/cooked_items.len)
+
+	for (var/obj/item/weapon/reagent_containers/food/snacks/S in contents)
+		S.cook()
+
+	eject(0) //clear out anything left
 
+	return
+
+/obj/machinery/microwave/process() // What you see here are the remains of proc/wzhzhzh, 2010 - 2019. RIP.
+	if (stat & (NOPOWER|BROKEN))
+		stop()
 		return
 
-/obj/machinery/microwave/proc/wzhzhzh(var/seconds as num) // Whoever named this proc is fucking literally Satan. ~ Z
-	for (var/i=1 to seconds)
-		if (stat & (NOPOWER|BROKEN))
-			return 0
-		use_power(active_power_usage)
-		sleep(10)
-	return 1
+	use_power(active_power_usage)
+
+	if(world.time > end_time)
+		stop()
+
+/obj/machinery/microwave/proc/half_time_process()
+	playsound(src, 'sound/machines/click.ogg', 20, 1)
+
+	if(failed)
+		visible_message(span("warning", "\The [src] begins to leak an acrid smoke..."))
 
 /obj/machinery/microwave/proc/has_extra_item()
 	for (var/obj/O in contents)
 		if ( \
 				!istype(O,/obj/item/weapon/reagent_containers/food) && \
 				!istype(O, /obj/item/weapon/grown) \
 			)
-			return 1
-	return 0
+			return TRUE
+	return FALSE
 
 /obj/machinery/microwave/proc/start()
-	src.visible_message("<span class='notice'>The microwave turns on.</span>", "<span class='notice'>You hear a microwave.</span>")
-	src.operating = 1
-	src.icon_state = "mw1"
-	src.updateUsrDialog()
+	start_time = world.time
+	end_time = cook_time + start_time
+	operating = TRUE
+
+	START_PROCESSING(SSprocessing, src)
+	addtimer(CALLBACK(src, .proc/half_time_process), cook_time / 2)

```

Comment:
If microwave is broken mid cooking timer just does it's thing anyway. Better migth want to add way to stop it or handle half time events inside `process()`